### PR TITLE
File Management APIs 

### PR DESCRIPTION
**Pull request by CraigLn**
Also additional of the UUID and Message Type to History
Setting the default History limit to be the max page size supported 
CryptoStream to encrypt/decrypt large items
Random IV (default to false for messages; true for files) support for Crypto functions

### CODE REVIEW

File: Tests/PubNubTests/Mocking/MockURLSession.swift [hunk 1]
```diff
@@ -71,60 +71,144 @@ class MockURLSessionTask: URLSessionDataTask {
...
+      guard let strongSelf = self else {
+        print("Failed to resumed due to weak self")
+        return
+      }
+      
+      if let data = task.mockData {
+        (strongSelf.delegate as? URLSessionDataDelegate)?.urlSession?(strongSelf.actualSession, dataTask: task, didReceive: data)
+      }
+      
+      (strongSelf.delegate as? URLSessionDataDelegate)?.urlSession?(strongSelf.actualSession, task: task, didCompleteWithError: task.error)
```
Comment (madisonsmith) [inline Tests/PubNubTests/Mocking/MockURLSession.swift:191]: Codacy found an issue: [Line should be 120 characters or less: currently 139 characters](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 9
           

Clones added
============
- Sources/PubNub/Helpers/XML/XMLEncoder.swift  12
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  3
- Sources/PubNub/Networking/Routers/ChannelGroupsRouter.swift  1
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
- Sources/PubNub/Models/PubNubFile.swift  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

File: Tests/PubNubTests/Networking/Routers/ChannelGroupEndpointTests.swift [hunk 2]
```diff
@@ -252,9 +252,10 @@ extension ChannelGroupsRouterTests {
     guard let sessions = try? MockURLSession.mockSession(for: ["maximumChannelCountExceeded_Message"]) else {
       return XCTFail("Could not create mock url session")
     }
+
```
Comment (madisonsmith) [inline Tests/PubNubTests/Networking/Routers/ChannelGroupEndpointTests.swift:255]: Codacy found an issue: [Lines should not have trailing whitespace.](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

File: Tests/PubNubTests/Mocking/MockURLSession.swift [hunk 3]
```diff
@@ -71,60 +71,144 @@ class MockURLSessionTask: URLSessionDataTask {
...
 
-      if let mockTask = self.responseForTask?(task, taskIndex) {
+      if let mockTask = strongSelf.responseForDataTask?(task, taskIndex) {
         mockTask.mockState = .completed
         if let data = mockTask.mockData {
-          self.dataDelegate?.urlSession?(self.actualSession, dataTask: task, didReceive: data)
+          strongSelf.dataDelegate?.urlSession?(strongSelf.actualSession, dataTask: task, didReceive: data)
         }
-        self.dataDelegate?.urlSession?(self.actualSession, task: mockTask, didCompleteWithError: mockTask.error)
+        strongSelf.dataDelegate?.urlSession?(strongSelf.actualSession, task: mockTask, didCompleteWithError: mockTask.error)
```
Comment (madisonsmith) [inline Tests/PubNubTests/Mocking/MockURLSession.swift:175]: Codacy found an issue: [Line should be 120 characters or less: currently 124 characters](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

File: Tests/PubNubTests/Mocking/MockURLSession.swift [hunk 4]
```diff
@@ -71,60 +71,144 @@ class MockURLSessionTask: URLSessionDataTask {
   }
 }
 
+class MockURLSessionUploadTask: URLSessionUploadTask {
+  var mockSession: URLSessionReplaceable
+
```
Comment (madisonsmith) [inline Tests/PubNubTests/Mocking/MockURLSession.swift:76]: Codacy found an issue: [Lines should not have trailing whitespace.](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

File: Tests/PubNubTests/Mocking/MockURLSession.swift [hunk 5]
```diff
@@ -133,6 +217,34 @@ class MockURLSession: URLSessionReplaceable {
     return task
   }
 
+  func uploadTask(withStreamedRequest request: URLRequest) -> URLSessionUploadTask {
+    guard let task = self.state.lockedRead({ $0.tasks.first { $0.originalRequest == request } }) as? URLSessionUploadTask else {
```
Comment (madisonsmith) [inline Tests/PubNubTests/Mocking/MockURLSession.swift:221]: Codacy found an issue: [Line should be 120 characters or less: currently 128 characters](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

File: Tests/PubNubTests/Mocking/MockURLSession.swift [hunk 6]
```diff
@@ -133,6 +217,34 @@ class MockURLSession: URLSessionReplaceable {
...
+
+  func uploadTask(with request: URLRequest, fromFile url: URL) -> URLSessionUploadTask {
+    guard let task = self.state.lockedRead({ $0.tasks.first { $0.originalRequest == request } }) as? URLSessionUploadTask else {
+      fatalError("Task not found for matching request \(request)")
+    }
+    return task
+  }
+
+  func downloadTask(with url: URL) -> URLSessionDownloadTask {
+    guard let task = self.state.lockedRead({ $0.tasks.first { $0.originalRequest?.url == url } }) as? URLSessionDownloadTask else {
```
Comment (madisonsmith) [inline Tests/PubNubTests/Mocking/MockURLSession.swift:235]: Codacy found an issue: [Line should be 120 characters or less: currently 131 characters](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

File: Tests/PubNubTests/Mocking/MockURLSession.swift [hunk 7]
```diff
@@ -71,60 +71,144 @@ class MockURLSessionTask: URLSessionDataTask {
...
+  
+  func resume(task: MockURLSessionUploadTask) {
+    delegateQueue.addOperation { [weak self] in
+      guard let strongSelf = self else {
+        print("Failed to resumed due to weak self")
+        return
+      }
+      
+      if let data = task.mockData {
+        (strongSelf.delegate as? URLSessionDataDelegate)?.urlSession?(strongSelf.actualSession, dataTask: task, didReceive: data)
```
Comment (madisonsmith) [inline Tests/PubNubTests/Mocking/MockURLSession.swift:188]: Codacy found an issue: [Line should be 120 characters or less: currently 129 characters](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

File: Tests/PubNubTests/Mocking/MockURLSession.swift [hunk 8]
```diff
@@ -133,6 +217,34 @@ class MockURLSession: URLSessionReplaceable {
...
+
+  func downloadTask(with url: URL) -> URLSessionDownloadTask {
+    guard let task = self.state.lockedRead({ $0.tasks.first { $0.originalRequest?.url == url } }) as? URLSessionDownloadTask else {
+      fatalError("Task not found for matching request \(url)")
+    }
+    return task
+  }
+
+  func downloadTask(withResumeData resumeData: Data) -> URLSessionDownloadTask {
+    guard let task = self.state.lockedRead({ $0.tasks.first { ($0 as? MockURLSessionDownloadTask)?.resumeData == resumeData } }) as? URLSessionDownloadTask else {
```
Comment (madisonsmith) [inline Tests/PubNubTests/Mocking/MockURLSession.swift:242]: Codacy found an issue: [Line should be 120 characters or less: currently 162 characters](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

File: Tests/PubNubTests/Mocking/MockURLSession.swift [hunk 9]
```diff
@@ -71,60 +71,144 @@ class MockURLSessionTask: URLSessionDataTask {
...
+  
+  func resume(task: MockURLSessionDownloadTask) {
+    delegateQueue.addOperation { [weak self] in
+      guard let strongSelf = self else {
+        print("Failed to resumed due to weak self")
+        return
+      }
+      
+      if let location = task.mockDownloadLocation {
+        (strongSelf.delegate as? URLSessionDownloadDelegate)?.urlSession(strongSelf.actualSession, downloadTask: task, didFinishDownloadingTo: location)
```
Comment (madisonsmith) [inline Tests/PubNubTests/Mocking/MockURLSession.swift:203]: Codacy found an issue: [Line should be 120 characters or less: currently 152 characters](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  2
- Sources/PubNub/Models/PubNubFile.swift  2
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  2
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
- Sources/PubNub/Models/PubNubFile.swift  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones added
============
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
- Sources/PubNub/Models/PubNubFile.swift  1
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  1
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

File: Examples/Sources/File/LocalFileExample.swift [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+import Foundation
+
+import PubNub
+
+enum FileError: Error {
+  case fileNameError
+  case missingLocalFile
+}
+
+struct LocalFileExample: PubNubLocalFile, Hashable  {
```
Comment (madisonsmith) [inline Examples/Sources/File/LocalFileExample.swift:37]: Codacy found an issue: [Opening braces should be preceded by a single space and on the same line as the declaration.](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

File: Examples/Sources/File/LocalFileExample.swift [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    fileURL = url
+    self.channel = channel
+
+    let filenameSplit = url.lastPathComponent.split(separator: ":")
+
+    guard let fileIdSubstring = filenameSplit.first,
+      let filenameSubstring = filenameSplit.last else {
+      throw FileError.fileNameError
+    }
+
```
Comment (madisonsmith) [inline Examples/Sources/File/LocalFileExample.swift:71]: Codacy found an issue: [Lines should not have trailing whitespace.](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  1
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
- Sources/PubNub/Models/PubNubFile.swift  1
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/PubNub/Helpers/Streams/CryptoInputStream.swift  2
- Sources/PubNub/Models/PubNubFile.swift  1
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  1
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  1
- Sources/PubNub/Helpers/Streams/CryptoInputStream.swift  2
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
- Sources/PubNub/Models/PubNubFile.swift  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  1
- Sources/PubNub/Helpers/Streams/CryptoInputStream.swift  2
- Tests/PubNubTests/Networking/Routers/FileManagementRouterTests.swift  41
- Sources/PubNub/Models/PubNubFile.swift  1
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
- Tests/PubNubTests/Mocking/MockURLSession.swift  5
- Tests/PubNubTests/Networking/Routers/PublishRouterTests.swift  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

File: Tests/PubNubTests/Mocking/MockURLSession.swift [hunk 12]
```diff
@@ -59,40 +59,187 @@ class MockURLSessionTask: URLSessionDataTask {
...
+  }
+
+  override var taskIdentifier: Int {
+    return mockIdentifier
+  }
+
+  override var countOfBytesExpectedToReceive: Int64 {
+    return 0
+  }
+
```
Comment (madisonsmith) [inline Tests/PubNubTests/Mocking/MockURLSession.swift:112]: Codacy found an issue: [Lines should not have trailing whitespace.](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
- Tests/PubNubTests/Networking/Routers/PublishRouterTests.swift  8
- Sources/PubNub/Helpers/Streams/CryptoInputStream.swift  2
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  1
- Sources/PubNub/Models/PubNubFile.swift  1
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
- Tests/PubNubTests/Mocking/MockURLSession.swift  2
- Tests/PubNubTests/Networking/Routers/FileManagementRouterTests.swift  41
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Tests/PubNubTests/Networking/Routers/FileManagementRouterTests.swift  41
- Tests/PubNubTests/Mocking/MockURLSession.swift  2
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  1
- Sources/PubNub/Helpers/Streams/CryptoInputStream.swift  2
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
- Sources/PubNub/Models/PubNubFile.swift  1
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
- Tests/PubNubTests/Networking/Routers/PublishRouterTests.swift  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

**Comment by CraigLn** [comment]
This should be good to go once the review threshold is met

File: Sources/PubNub/Networking/HTTPFileTask.swift [hunk 13]
```diff
@@ -0,0 +1,471 @@
...
+      return
+    }
+
+    let temporaryURL = fileDownloadTask.temporaryURL()
+
+    do {
+      // Move file to a temporary location or it's lost at the end of this scope
+      try FileManager.default.moveItem(at: location, to: temporaryURL)
+
+      didDownload?(session, downloadTask, temporaryURL)
```
Comment (parfeon) [inline Sources/PubNub/Networking/HTTPFileTask.swift:439]: @CraigLn what if we will add same behaviour as `downloadTask(with:completionHandler:)` has (thought, there is completion closure instead of delegate callbacks) where file will be removed from temporary location right after closure call will be completed?

**Review by parfeon — COMMENTED** [review]
I've left some comments (mostly because of parameters which I didn't sae in specs).

File: Tests/PubNubTests/Mocking/Responses/File/file_generateURL_success.json [hunk 14]
```diff
@@ -0,0 +1,53 @@
+{
```
[the remark this answers is not in the document]
Comment (CraigLn) [reply Tests/PubNubTests/Mocking/Responses/File/file_generateURL_success.json:1]: Good catch; will change as the values aren't very important for testing.

File: Sources/PubNub/Networking/Routers/FileManagementRouter.swift [hunk 15]
```diff
@@ -0,0 +1,538 @@
...
+  }
+
+  var channel: String
+  var createdDate: Date?
+
+  var concreteMessage: AnyJSON?
+  var message: JSONCodable? {
+    get { return concreteCustom }
+    set { concreteCustom = newValue?.codableValue }
+  }
```
[the remark this answers is not in the document]
Comment (CraigLn) [reply Sources/PubNub/Networking/Routers/FileManagementRouter.swift:442]: Just checked, and it's not used.  Updated to verify with tests that the publish payload is correctly matching the expected payload.

File: Sources/PubNub/PubNub.swift [hunk 16]
```diff
@@ -1097,29 +1115,40 @@ extension PubNub {
...
   ) {
     let router: HistoryRouter
-    if actions {
+    if includeActions {
       router = HistoryRouter(
-        .fetchWithActions(channel: channels.first ?? "",
-                          max: page?.limit, start: page?.start, end: page?.end, includeMeta: includeMeta),
+        .fetchWithActions(
+          channel: channels.first ?? "",
+          max: page?.limit ?? 100, start: page?.start, end: page?.end,
```
[the remark this answers is not in the document]
Comment (CraigLn) [reply Sources/PubNub/PubNub.swift:1138]: Updating

File: Sources/PubNub/Networking/Routers/HistoryRouter.swift [hunk 17]
```diff
@@ -32,8 +32,14 @@ import Foundation
 struct HistoryRouter: HTTPRouter {
   // Nested Endpoint
   enum Endpoint: CustomStringConvertible {
-    case fetch(channels: [String], max: Int?, start: Timetoken?, end: Timetoken?, includeMeta: Bool)
-    case fetchWithActions(channel: String, max: Int?, start: Timetoken?, end: Timetoken?, includeMeta: Bool)
+    case fetch(
+      channels: [String], max: Int?, start: Timetoken?, end: Timetoken?,
+      includeMeta: Bool, includeMessageType: Bool, includeUUID: Bool
```
[the remark this answers is not in the document]
Comment (CraigLn) [reply Sources/PubNub/Networking/Routers/HistoryRouter.swift:37]: The decision was previously made to try and make the router best match the requirements from the server, and the PubNub layer match the business case defaults set by the PMs.

tl;dr Server has them as optionals that default to something, but PMs want it to be explicit.  This reflects both of those.

File: Sources/PubNub/Networking/Routers/FileManagementRouter.swift [hunk 18]
```diff
@@ -0,0 +1,538 @@
...
+    case name
+    case size
+    case mimeType = "mime-type"
+    case custom
+  }
+
+  init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+    concreteMessage = try container.decodeIfPresent(AnyJSON.self, forKey: .message)
+    channel = try container.decodeIfPresent(String.self, forKey: .channel) ?? ""
```
[the remark this answers is not in the document]
Comment (CraigLn) [reply Sources/PubNub/Networking/Routers/FileManagementRouter.swift:493]: It's omitted from this decoder, but the protocol requires a channel to be tightly coupled with the channel.  The encoder for this object will leave out the channel to reflect the expectation of other SDKs.

File: Sources/PubNub/Networking/Routers/FileManagementRouter.swift [hunk 19]
```diff
@@ -0,0 +1,538 @@
...
+  func encode(to encoder: Encoder) throws {
+    var rootContainer = encoder.container(keyedBy: CodingKeys.self)
+    try rootContainer.encodeIfPresent(concreteMessage, forKey: .message)
+
+    var fileContainer = rootContainer.nestedContainer(keyedBy: FileCodingKeys.self, forKey: .file)
+    try fileContainer.encode(fileId, forKey: .id)
+    try fileContainer.encode(filename, forKey: .name)
+    try fileContainer.encode(size, forKey: .size)
+    try fileContainer.encode(contentType, forKey: .mimeType)
+    try fileContainer.encodeIfPresent(concreteCustom, forKey: .custom)
```
[the remark this answers is not in the document]
Comment (CraigLn) [reply Sources/PubNub/Networking/Routers/FileManagementRouter.swift:512]: See above response

File: Sources/PubNub/Networking/Routers/ChannelGroupsRouter.swift [hunk 20]
```diff
@@ -191,6 +191,8 @@ enum QueryKey: String {
   case count
   case max
   case includeMeta = "include_meta"
+  case includeMessageType = "include_message_type"
+  case includeUUID = "include_uuid"
```
[the remark this answers is not in the document]
Comment (CraigLn) [reply Sources/PubNub/Networking/Routers/ChannelGroupsRouter.swift:195]: No I think they just somehow got put there and weren't moved to the HTTPRouter file.

File: Sources/PubNub/Networking/HTTPFileTask.swift (hunk 13 above)
Comment (CraigLn) [reply Sources/PubNub/Networking/HTTPFileTask.swift:439]: I'm not sure.  I only tested how it performs inside the delegate, but can confirm that the URL contents will be deleted at the end of the scope unless moved (even if it's to another location in the `tmp` directory).

File: Sources/PubNub/PubNub.swift [hunk 21]
```diff
@@ -1097,29 +1115,40 @@ extension PubNub {
...
+          includeMeta: includeMeta, includeMessageType: includeMessageType,
+          includeUUID: includeUUID
+        ),
         configuration: configuration
       )
     } else {
       router = HistoryRouter(
-        .fetch(channels: channels, max: page?.limit, start: page?.start, end: page?.end, includeMeta: includeMeta),
+        .fetch(
+          channels: channels, max: page?.limit ?? 100,
```
[the remark this answers is not in the document]
Comment (CraigLn) [reply Sources/PubNub/PubNub.swift:1147]: See previous comment.

File: Sources/PubNub/Networking/Routers/FileManagementRouter.swift [hunk 22]
```diff
@@ -0,0 +1,538 @@
...
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+    concreteMessage = try container.decodeIfPresent(AnyJSON.self, forKey: .message)
+    channel = try container.decodeIfPresent(String.self, forKey: .channel) ?? ""
+
+    let fileContainer = try container.nestedContainer(keyedBy: FileCodingKeys.self, forKey: .file)
+    fileId = try fileContainer.decode(String.self, forKey: .id)
+    filename = try fileContainer.decode(String.self, forKey: .name)
+    size = try fileContainer.decode(Int64.self, forKey: .size)
+    contentType = try fileContainer.decode(String.self, forKey: .mimeType)
+    concreteCustom = try fileContainer.decodeIfPresent(AnyJSON.self, forKey: .custom)
```
[the remark this answers is not in the document]
Comment (CraigLn) [reply Sources/PubNub/Networking/Routers/FileManagementRouter.swift:500]: It's not on the spec, but the goal was to provide a customization base to the file objects to be more robust.  I was hesitant to include it here, and will move it to the framework layer.

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  1
- Sources/PubNub/Models/PubNubFile.swift  1
- Tests/PubNubTests/Mocking/MockURLSession.swift  5
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
- Tests/PubNubTests/Networking/Routers/FileManagementRouterTests.swift  41
- Sources/PubNub/Helpers/Streams/CryptoInputStream.swift  2
- Tests/PubNubTests/Networking/Routers/PublishRouterTests.swift  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

File: Sources/PubNub/Networking/HTTPFileTask.swift (hunk 13 above)
Comment (parfeon) [reply Sources/PubNub/Networking/HTTPFileTask.swift:439]: @CraigLn it is fine. I thought that there can be constraints exactly because it is delegate callbacks.
Than I'm fine with current approach (just wanted some input). 
Thanks!

File: Sources/PubNub/Networking/Routers/ChannelGroupsRouter.swift (hunk 20 above)
[the remark this answers is not in the document]
Comment (parfeon) [reply Sources/PubNub/Networking/Routers/ChannelGroupsRouter.swift]: I see they have been moved to proper location :)

File: Sources/PubNub/Networking/Routers/FileManagementRouter.swift (hunk 22 above)
[the remark this answers is not in the document]
Comment (parfeon) [reply Sources/PubNub/Networking/Routers/FileManagementRouter.swift]: @CraigLn let's not forget, that this payload will arrive not only to Swift SDK. If there ever will appear another SDK, which map response to models, it may experience some issue, because Swift will include unexpected field in payload.

File: Sources/PubNub/Networking/Routers/FileManagementRouter.swift [hunk 23]
```diff
@@ -0,0 +1,538 @@
...
+struct FilePublishPayload: PubNubFile {
+  /// The additional payload published along with the file
+  var additionalDetails: JSONCodable? {
+    get {
+      return additionalDetailsConcrete
+    }
+    set {
+      additionalDetailsConcrete = newValue?.codableValue
+    }
+  }
```
[the remark this answers is not in the document]
Comment (parfeon) [reply Sources/PubNub/Networking/Routers/FileManagementRouter.swift:416]: @CraigLn is there any input on this?

File: Sources/PubNub/Networking/Routers/FileManagementRouter.swift [hunk 24]
```diff
@@ -0,0 +1,538 @@
...
+    case file
+    case channel
+  }
+
+  enum FileCodingKeys: String, CodingKey {
+    case id
+    case name
+    case size
+    case mimeType = "mime-type"
+    case custom
```
[the remark this answers is not in the document]
Comment (parfeon) [reply Sources/PubNub/Networking/Routers/FileManagementRouter.swift:486]: @CraigLn `custom` field probably is part of discussion below.

File: Sources/PubNub/Networking/Routers/FileManagementRouter.swift (hunk 22 above)
[the remark this answers is not in the document]
Comment (CraigLn) [reply Sources/PubNub/Networking/Routers/FileManagementRouter.swift]: Agreed, and it has been removed.

File: Sources/PubNub/Networking/Routers/FileManagementRouter.swift (hunk 24 above)
[the remark this answers is not in the document]
Comment (CraigLn) [reply Sources/PubNub/Networking/Routers/FileManagementRouter.swift]: Removed

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Tests/PubNubTests/Mocking/MockURLSession.swift  5
- Tests/PubNubTests/Networking/Routers/FileManagementRouterTests.swift  41
- Tests/PubNubTests/Networking/Routers/HistoryRouterTests.swift  4
- Tests/PubNubTests/Networking/Routers/PublishRouterTests.swift  8
- Sources/PubNub/Networking/Routers/FileManagementRouter.swift  1
- Sources/PubNub/Helpers/Streams/CryptoInputStream.swift  2
- Sources/PubNub/Helpers/XML/XMLDecoder.swift  85
- Sources/PubNub/Models/PubNubFile.swift  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=6302911&bid=20497439)

**Review by parfeon — APPROVED** [review]
LGTM
